### ISSUE TITLE
Annotation to support singleton pattern.

### ISSUE THREAD

**Issue by rzvc** [vimeo/psalm#7952]
I can't find a way to indicate to Psalm that a method/function will always return the same object:
```php
class Foo
{
    private static ?Foo $instance = null;
    public ?string $str = 'asdf';

    private function __construct()
    {
    }

    public static function getInstance() : self
    {
        if (is_null(self::$instance))
                self::$instance = new Foo();

        return self::$instance;
    }
}

if (is_null(Foo::getInstance()->str))
    return;

echo trim(Foo::getInstance()->str);     // Complains about $str possibly being null.


$foo = Foo::getInstance();

if (is_null($foo->str))
    return;

echo trim($foo->str);     // As expected, it doesn't complain here.
```
https://psalm.dev/r/55b5ff1a69

Something like ```@psalm-singleton``` on the method's or function's docblock would be nice.

**Comment by orklah** [comment]
Even if Psalm understands it's dealing with a singleton, it won't help you. 

Any statement between your is_null check and your trim could potentially mutate your instance and change $str value...

**Comment by rzvc** [comment]
That can be said about any object, and it would apply to the scenario where the object is used directly as well, but that one works. If Psalm would be aware of the singleton pattern, it would behave the same in both cases.
